feat(Options): customize loading indicator with options

NathanWalker opened the pull request:
This is a preview at this point to see where I'm headed and get your feedback.
I need to do more Android testing, but this works great on iOS so far :+1:

NathanWalker committed:
feat(Options): customize loading indicator with options

NathanWalker commented:
Ok I have a working Android demo now.
I squashed the commits and all is in good shape.

maknz commented:
Thanks for your work man. Will take a gander tomorrow :)

On Mon, Feb 1, 2016 at 7:57 PM, Nathan Walker notifications@github.com
wrote:

> Ok I have a working Android demo now.
> 
> ## I squashed the commits and all is in good shape.
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/pocketsmith/nativescript-loading-indicator/pull/3#issuecomment-177814390

NathanWalker commented:
I can help with documentation in the README after you've had chance to look it all over, n/p.

maknz commented on demo/app/main-page-android.xml:
```diff
@@ -0,0 +1,11 @@
+  <GridLayout rows="*,auto">
+    <ScrollView>
+      <StackLayout row="0">
+        <Button text="Message" tap="{{showLoader}}" />
```
Could we have a space either side so this would read `{{ showLoader }}` please? Likewise for the other bindings :smile:.

maknz commented on demo/app/main-view-model-android.ts:
```diff
@@ -0,0 +1,42 @@
+  
+  private demoProgress() {
+    setTimeout(() => {
+      loadingIndicator.show({progress: 15});
```
Could we space these out like the above so it would read `loadingIndicator.show({ progress: 15 });`

maknz commented on demo/app/main-view-model-android.ts:
```diff
@@ -0,0 +1,42 @@
+export class LoadingAndroidTest extends Observable {
+  
+  public showLoader() {
+    loadingIndicator.show({ message: 'Loading test...' });
```
Small thing, but could we use double quotes for strings throughout the project? This is a {N} convention that we're trying to keep to.

NathanWalker replied:
Sure. I will mention though with TypeScript I believe single quotes are preferred, and a common convention is also using backticks in implementation files so whenever the need arises to interpolate, you can:

```
let opinion = 'awesome';
`this is ${opinion}`;
```

maknz commented on loading-indicator.ios.js:
```diff
@@ -1,11 +1,46 @@
+  if (options.margin) this._hud.margin = options.margin;
+  if (options.dimBackground) this._hud.dimBackground = true;
+  if (options.color) {
+    this._hud.color = UIColor.colorWithRedGreenBlueAlpha(options.color.r, options.color.g, options.color.b, options.color.a);
```
Could we possibly use {N}'s Color utility for this?

``` js
var Color = require("color").Color;

this._hud.color = new Color("#hex").ios;
```

NathanWalker replied:
Great feedback! Will update PR when I can.

maknz commented:
Overall this is looking pretty good. A couple of things:

It seems only `message` and `progress` are cross-platform. Everything else is specific to iOS and Android. I think what would be good is that any iOS-only preferences go in an `ios` embedded object, and likewise for Android. For example:

``` js
loader.show({
  message: "Loading, please wait...",
  ios: {
    color: "#abcdef",
    dimBackground: true
  },
  android: {
    progressStyle: "something"
  }
});
```

The other thing is `mode` for MBProgressHUD. I don't think we should use numbers here since the user won't know what that means. We should either use enums/constants in the `loadingIndicator` module, or just let them pass the `MBProgressHUD` constants through, e.g.

``` js
loader.show({
  message: "Hello!",
  ios: {
    mode: MBProgressHUDModeCustomView,
    customView: "something.jpg"
  }
});
```

@lscown is gonna give this a run shortly :-)

NathanWalker commented:
Thanks @maknz really good feedback. I updated the PR, squashed commits to fix a few of the comments. It's all I had time for at the moment, but feel free to pull:
https://github.com/NathanWalker/nativescript-loading-indicator/tree/support-options
and tweak further!

lscown commented:
Hey @NathanWalker! 

The demo works great, both iOS and Android (APIs 17 -> 23). It'd be great to see more options on the Android side, especially a determinate loader. Other than that, everything looks all good :-)

NathanWalker commented:
I'm getting back into this PR today. I'll use platform specific options for those that are as mentioned above:

```
loader.show({
  message: "Loading, please wait...",
  ios: {
    color: "#abcdef",
    dimBackground: true
  },
  android: {
    progressStyle: "something"
  }
});
```

Also, would you guys mind if I convert the plugin code to TypeScript?  The demo is in TS, and it would be nice to have this plugin's source be TS as well.

maknz commented:
I'm not opposed to it, but we aren't set up for it. What's your process like for compiling and running the plugin?

NathanWalker commented:
@maknz I wouldn't submit a PR converting to TS without getting you guys all setup silly pants ;)

Just pushed the conversion.
This sets you up completely with fully converted TypeScript source and a build workflow to publish.
I just wrote an article about this which will be published on Telerik Network tonight/tomorrow but to give you the gist:
- `.js` files don't need to be on Github
- You run `npm run build` before you publish to generate all *.js files
- The plugin is published with `.js` files to be consumable
- keeps github clean and npm very clean

I'll link the full article once it goes live, but this gets you setup properly.

I'm more an iOS developer than Android so if you want to build out the Android demo more, feel free post-merge. Thanks.

NathanWalker commented on tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+        "sourceMap": true
+    },
+    "files": [
+      "demo/node_modules/tns-core-modules/tns-core-modules.d.ts",
```
We may remove this. It depends. It's definitely for smooth developer workflow only. Means nothing to consumers. It helps remove all warnings when building.
However, for first time setup (when a developer has not `npm install` inside the demo yet), it would not be available yet. Let me know what you guys think. I could go either way with this.

NathanWalker commented:
Here's the article:
http://developer.telerik.com/featured/creating-nativescript-plugins-in-typescript/

maknz commented on loading-indicator.ios.ts:
```diff
@@ -0,0 +1,50 @@
+declare var MBProgressHUDModeCustomView: any;
+declare var UIApplication: any;
+declare var UIImageView: any;
+declare var UIImage: any;
```
Are these needed such that we don't get warnings from typescript not knowing about those constants?

NathanWalker replied:
Yes sir, those help avoid the TS warnings.

maknz commented on loading-indicator.android.ts:
```diff
@@ -0,0 +1,49 @@
+        let cancelable = false;
+        if (options.android) {
+          if (typeof options.android.indeterminate !== 'undefined') indeterminate = options.android.indeterminate;
+          if (typeof options.android.cancelable !== 'undefined') cancelable = options.android.cancelable;
```
Is there a reason for this more verbose check for the field being provided, compared to the simpler ones below? I think I like the simpler ones better, feels like it provides the same safety anyway.

And you could also do:

``` js
if (options.android.cancelable !== undefined)
```

I think?

NathanWalker replied:
I alway side on being very explicit with stuff like that, however you may be correct, I'll try it your way and if all works out, I'll update PR.

maknz commented on loading-indicator.common.ts:
```diff
@@ -0,0 +1,26 @@
+interface AndroidOptionsI {
```
Just curious: is the `I` to denote an interface? IMO I'd prefer without the I and it's up to the user to understand it's an interface and not a class they can use. I think this is how NativeScript do it internally. What are your thoughts on that?

NathanWalker replied:
The `I` is somewhat typical of an interface. I kind of prefer it. If big deal to you, I can change it. Not 100% sure what core team does but I've seen it in various developments. I think it does help the understanding.

maknz commented:
@NathanWalker overall I'm pretty happy with this. If we can discuss and resolve the above, I think we'll be good to merge and publish. We'll have to bump up to 2.0.0 due to the breaking change around needing to create a `new LoadingIndicator()`. It's definitely a better design though, peachy.

NathanWalker commented:
@maknz @lscown Ok got the adjustments made. Interfaces renamed without 'I' and adjusted the `undefined` concern.
Also I organized things a little better here. This allows the plugin for both platforms `android` and `ios` to export the interfaces. This will allow consumers to use the interfaces to provide intellisense around what options are available for each :+1: 

Lastly, I added a `CONTRIBUTING.md` doc which documents how to run the demo as well as a suggested development workflow with helper npm scripts. It's a process I use myself and it works really well. When you get it merged, you should try this workflow. (_You may need to tweak the `demo.android` task to include better command line options for what you use_, The `demo.ios` script works great for the Simulator).

You can expand on the android demo to flex more of the options but that can be done down the road.

If this could be merged, version bumped, and released would be great. I'm looking forward to using it in one of my projects when you do! Thanks!

maknz commented:
All looking good @NathanWalker. Thanks for your work on this.
